ts/Stock/Indicators/DisparityIndex/DisparityIndexIndicator.ts

```diff
@@ -0,0 +1,282 @@
+             *
+             * If value is different than ema|dema|tema|wma, then sma is used.
+             */
+            average: 'sma',
+            index: 3,
+            period: 14
+        },
+        marker: {
+            enabled: false
+        },
+        dataGrouping: {
+            approximation: 'averages'
+        }
+    } as DisparityIndexOptions);
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    public data: Array<DisparityIndexPoint> = void 0 as any;
+    public options: DisparityIndexOptions = void 0 as any;
+    public points: Array<DisparityIndexPoint> = void 0 as any;
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    public init(this: DisparityIndexIndicator): void {
+        const args = arguments,
+            ctx = this, // Disparity Index indicator
+            params = args[1].params, // options.params
+            averageType = params && params.average ? params.average : void 0;
+
+        let averageIndicator;
+
+        switch (averageType) {
+        case 'ema':
+            averageIndicator = EMAIndicator;
+            break;
+        case 'dema':
+            averageIndicator = DEMAIndicator;
+            break;
+        case 'tema':
+            averageIndicator = TEMAIndicator;
+            break;
+        case 'wma':
+            averageIndicator = WMAIndicator;
+            break;
+        default: // use sma if any of the above strings do not match
+            averageIndicator = SMAIndicator;
+        }
+
+        ctx.averageIndicator = averageIndicator;
+
+        if (averageType !== 'sma') { // no need to check whether sma is loaded
+            // Check if the required average indicator modules is loaded
+            RequiredIndicatorMixin.isParentLoaded(
+                averageIndicator as any,
+                averageType,
+                ctx.type,
+                function (indicator: Highcharts.Indicator): undefined {
+                    indicator.prototype.init.apply(ctx, args);
+                    return;
+                }
+            );
+        }
+    }
+
+    public calculateDisparityIndex(
+        close: number,
+        periodAverage: number
+    ): number {
+        return correctFloat(close - periodAverage) / periodAverage * 100;
+    }
+
+    public getValues<TLinkedSeries extends LineSeries>(
+        series: TLinkedSeries,
+        params: DisparityIndexParamsOptions
+    ): (IndicatorValuesObject<TLinkedSeries>|undefined) {
+        const averageType = params.average,
+            index = params.index,
+            period = params.period as any,
+            xVal: Array<number> = (series.xData as any),
+            yVal: Array<number>|Array<Array<number>> = (series.yData as any),
+            yValLen: number = yVal ? yVal.length : 0,
+            disparityIndexPoint: Array<Array<number>> = [],
+            xData: Array<number> = [],
+            yData: Array<number> = [],
+            averageIndicator: any = this.averageIndicator,
```

Comment:
I think you might improve the types here.

Reply:
As I think about it more thoroughly it might not be that easy. In the beginning, I thought that it's only the string, but I think we should accept `any` here.

Reply:
Could you elaborate? `averageIndicator` can be either undefined or one of the predefined other Indicator types, right?